File: Find all distinct subset (or subsequence) sums - GFG/Find all distinct subset (or subsequence) sums.cpp
// two solutions:


// 1.
class Solution {
public:
    set<int> s;
    int dp[101][10001];
	vector<int> DistinctSum(vector<int>nums){
	    int n = nums.size();
	    memset(dp, -1, sizeof(dp));
	    solve(0, nums, n);
	    vector<int> ans;
	    for(auto& val: s) {
	        ans.push_back(val);
	    }
	    return ans;
	}
	void solve(int sum, vector<int>& nums, int n) {
	    if(n==0) {
	        s.insert(sum);
	        return;
	    }
	    if(dp[n][sum]==1) return;
	    solve(sum+nums[n-1], nums, n-1);
	    solve(sum, nums, n-1);
	    dp[n][sum] = 1;
	}
};


// 2. 
class Solution {
public:
	vector<int> DistinctSum(vector<int> nums){
	    // Code here
	    set<int> st;
	    st.insert(0);
	    for(auto num : nums){
	        vector<int> v;
	        for(auto i : st) v.push_back(i + num);
	        for(auto vv : v) st.insert(vv);
	    }
	    nums.clear();
	    for(auto it : st) nums.push_back(it);
	    return nums;
	}
};

File: Find all distinct subset (or subsequence) sums.java
class Solution
{
    public int[] DistinctSum(int[] nums)
    {
        // Code here
        boolean[][]dp=new boolean[101][10001];
        Set<Integer> set=new HashSet<>();
        dfs(nums,0,0,set,dp);
        int[] ans=new int[set.size()];
        int i=0;
        for(int ele:set){
            ans[i]=ele;
            i++;
        }
        Arrays.sort(ans);
        return ans;
    }
    void dfs(int[]nums,int i,int sum,Set<Integer> set,boolean[][]dp){
        if(i==nums.length){
            set.add(sum);
            return;
        }
        if(dp[i][sum])return;
        dp[i][sum]=true;
        dfs(nums,i+1,sum,set,dp);
        dfs(nums,i+1,sum+nums[i],set,dp);
    }
}
